cumulus/polkadot-omni-node/lib/src/command.rs

```diff
@@ -100,7 +102,47 @@ fn new_node_spec(
 
 /// Parse command line arguments into service configuration.
 pub fn run<CliConfig: crate::cli::CliConfig>(cmd_config: RunConfig) -> Result<()> {
-	let mut cli = Cli::<CliConfig>::from_args();
+	run_with_custom_cli::<CliConfig, DefaultExtraSubcommands>(cmd_config)
+}
+
+/// Parse command‑line arguments into service configuration and inject an
+/// optional extra sub‑command.
+///
+/// `run_with_custom_cli` builds the base CLI for the node binary, then asks the
+/// `Extra` type for an optional extra sub‑command via ExtraSubcommand
+///
+/// When the user actually invokes that extra sub‑command,
+/// `Extra::from_arg_matches` returns a parsed value which is immediately passed
+/// to `extra.handle(&cfg)` and the process exits.  Otherwise control falls
+/// through to the normal node‑startup / utility sub‑command match.
+///
+/// # Type Parameters
+/// * `CliConfig` – customization trait supplying user‑facing info (name, description, version) for
+///   the binary.
+/// * `Extra` – an implementation of [`ExtraSubcommand`].  Use *[`DefaultExtraSubcommands`]* in
+///   binaries that want some extra subcommands such as `export‑chain-spec`; use
+///   *[`NoExtraSubcommand`]* when the binary should expose no extra subcommands.
+pub fn run_with_custom_cli<CliConfig, Extra>(cmd_config: RunConfig) -> Result<()>
+where
+	CliConfig: crate::cli::CliConfig,
+	Extra: ExtraSubcommand,
+{
+	let cli_command = Cli::<CliConfig>::augment_args(<Extra as CommandFactory>::command());
+	let cli_command = Cli::<CliConfig>::setup_command(cli_command);
+	let matches = cli_command.get_matches();
+
+	let extra_has_subcommands = <Extra as CommandFactory>::command().has_subcommands();
```

Comment:
nit: (`Extra` should be a `CommandFactory` given the trait bounds)
```suggestion
	let extra_has_subcommands = Extra::command().has_subcommands();
```